src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -132347,20 +132375,385 @@ class A<T> where T : class?
+    static T? F8<U>(U? u) where U : T? => u;
+}";
+            comp = CreateCompilation(source6, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (6,41): warning CS8603: Possible null reference return.
+                //     static T F2<U>(U u) where U : T? => u; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "u").WithLocation(6, 41),
+                // (9,41): warning CS8603: Possible null reference return.
+                //     static T F5<U>(U? u) where U : T => u; // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "u").WithLocation(9, 41),
+                // (10,42): warning CS8603: Possible null reference return.
+                //     static T F6<U>(U? u) where U : T? => u; // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "u").WithLocation(10, 42));
         }
 
         [Fact]
+        [WorkItem(46150, "https://github.com/dotnet/roslyn/issues/46150")]
         public void UnconstrainedTypeParameter_28()
         {
             var source =
+@"#nullable disable
+class A<T, U> where U : T
+#nullable enable
+{
+    static T F1(U u) => u;
+    static T? F2(U u) => u;
+    static T F3(U? u) => u; // 1
+    static T? F4(U? u) => u;
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (7,26): warning CS8603: Possible null reference return.
+                //     static T F3(U? u) => u; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "u").WithLocation(7, 26));
+        }
+
+        [Fact]
+        [WorkItem(46150, "https://github.com/dotnet/roslyn/issues/46150")]
+        public void UnconstrainedTypeParameter_29()
+        {
+            var source1 =
+@"#nullable enable
+class A
+{
+    static object F1<T>(T t) => t; // 1
+    static object? F2<T>(T t) => t;
+    static object F3<T>(T? t) => t; // 2
+    static object? F4<T>(T? t) => t;
+}";
+            var comp = CreateCompilation(source1, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (4,33): warning CS8603: Possible null reference return.
+                //     static object F1<T>(T t) => t; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "t").WithLocation(4, 33),
+                // (6,34): warning CS8603: Possible null reference return.
+                //     static object F3<T>(T? t) => t; // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "t").WithLocation(6, 34));
+
+            var source2 =
+@"#nullable enable
+class A
+{
+    static object F1<T>(T t) where T : class => t;
+    static object? F2<T>(T t) where T : class => t;
+    static object F3<T>(T? t) where T : class => t; // 1
+    static object? F4<T>(T? t) where T : class => t;
+}";
+            comp = CreateCompilation(source2, parseOptions: TestOptions.RegularPreview);
```

Comment:
is the language version relevant? If not, consider using C# 8. 
Also applicable to other scenarios in this test.